File: server/livekitService.js
import { AccessToken } from 'livekit-server-sdk';
import jwt from 'jsonwebtoken';

export class LiveKitService {
  constructor() {
    this.apiKey = process.env.LIVEKIT_API_KEY;
    this.apiSecret = process.env.LIVEKIT_API_SECRET;
    this.wsUrl = process.env.LIVEKIT_WS_URL;
    
    console.log('[LiveKitService] Environment Variables:');
    console.log('- API_KEY:', this.apiKey ? 'Set' : 'Not set');
    console.log('- API_SECRET:', this.apiSecret ? 'Set' : 'Not set');
    console.log('- WS_URL:', this.wsUrl ? `"${this.wsUrl}"` : 'Not set');
    
    // Clean the WebSocket URL but don't invalidate it
    if (this.wsUrl) {
      this.wsUrl = this.wsUrl.trim().replace(/['"]/g, '');
      console.log('- WS_URL (cleaned):', `"${this.wsUrl}"`);
    }
    
    if (!this.apiKey || !this.apiSecret || !this.wsUrl) {
      console.warn('[LiveKitService] LiveKit credentials not configured properly. Voice interviews will be disabled.');
    } else {
      console.log('[LiveKitService] LiveKit service initialized successfully');
    }
  }

  isValidWebSocketUrl(url) {
    try {
      // Basic validation first
      if (!url || typeof url !== 'string') {
        console.warn('URL is not a valid string:', typeof url, url);
        return false;
      }
      
      const trimmedUrl = url.trim();
      console.log('Validating URL:', `"${trimmedUrl}"`);
      
      if (trimmedUrl === '') {
        console.warn('URL is empty after trimming');
        return false;
      }
      
      // Check if it starts with ws:// or wss://
      if (!trimmedUrl.startsWith('ws://') && !trimmedUrl.startsWith('wss://')) {
        console.warn('URL does not start with ws:// or wss://');
        return false;
      }
      
      // Parse URL to validate format
      let urlObj;
      try {
        urlObj = new URL(trimmedUrl);
      } catch (parseError) {
        console.warn('URL parsing failed:', parseError.message);
        return false;
      }
      
      // Validate protocol
      if (urlObj.protocol !== 'ws:' && urlObj.protocol !== 'wss:') {
        console.warn('Invalid protocol:', urlObj.protocol);
        return false;
      }
      
      // Validate hostname exists
      if (!urlObj.hostname) {
        console.warn('Invalid hostname:', urlObj.hostname);
        return false;
      }
      
      console.log('URL validation successful:', {
        protocol: urlObj.protocol,
        hostname: urlObj.hostname,
        port: urlObj.port
      });
      
      return true;
      
    } catch (error) {
      console.error('Unexpected error in URL validation:', error);
      return false;
    }
  }

  isConfigured() {
    const configured = !!(this.apiKey && this.apiSecret && this.wsUrl);
    console.log('[LiveKitService] Configuration check:', {
      apiKey: !!this.apiKey,
      apiSecret: !!this.apiSecret,
      wsUrl: !!this.wsUrl,
      configured
    });
    return configured;
  }

  getWebSocketUrl() {
    console.log('[LiveKitService] getWebSocketUrl() called, returning:', `"${this.wsUrl}"`);
    return this.wsUrl;
  }

  /**
   * Generate access token for a participant to join a LiveKit room
   */
  generateAccessToken(roomName, participantName, metadata = {}) {
    if (!this.isConfigured()) {
      console.error('[LiveKitService] Cannot generate token - not configured');
      throw new Error('LiveKit not configured');
    }

    console.log('[LiveKitService] Generating access token for:', {
      roomName,
      participantName,
      apiKeyLength: this.apiKey?.length || 0,
      apiSecretLength: this.apiSecret?.length || 0
    });

    try {
      // Validate API key format - LiveKit API keys should start with 'API'
      if (!this.apiKey.startsWith('API')) {
        console.error('[LiveKitService] Invalid API key format - should start with "API"');
        throw new Error('Invalid LiveKit API key format');
      }

      // Validate API secret length - should be a reasonable length
      if (this.apiSecret.length < 32) {
        console.error('[LiveKitService] API secret appears too short');
        throw new Error('Invalid LiveKit API secret - too short');
      }

      const at = new AccessToken(this.apiKey, this.apiSecret, {
        identity: participantName,
        name: participantName,
        metadata: JSON.stringify(metadata)
      });

      // Grant permissions for the participant
      at.addGrant({
        roomJoin: true,
        room: roomName,
        canPublish: true,
        canSubscribe: true,
        canPublishData: true,
        canUpdateOwnMetadata: true,
      });

      const token = at.toJwt();
      
      // Validate that the token is actually a string
      if (typeof token !== 'string' || token.length === 0) {
        console.error('[LiveKitService] Generated token is not a valid string:', {
          type: typeof token,
          length: token?.length || 0,
          token: token
        });
        throw new Error(`Invalid token generated: expected string, got ${typeof token} with length ${token?.length || 0}`);
      }
      
      console.log('[LiveKitService] Token generated successfully, length:', token.length);
      return token;
    } catch (error) {
      console.error('[LiveKitService] Error generating access token:', error);
      throw new Error(`Failed to generate access token: ${error.message}`);
    }
  }

  /**
   * Generate access token for the AI interviewer bot
   */
  generateInterviewerToken(roomName, interviewConfig) {
    if (!this.isConfigured()) {
      console.error('[LiveKitService] Cannot generate interviewer token - not configured');
      throw new Error('LiveKit not configured');
    }

    const interviewerIdentity = `ai-interviewer-${Date.now()}`;
    
    console.log('[LiveKitService] Generating interviewer token for:', {
      roomName,
      interviewerIdentity
    });

    try {
      // Validate API key format - LiveKit API keys should start with 'API'
      if (!this.apiKey.startsWith('API')) {
        console.error('[LiveKitService] Invalid API key format - should start with "API"');
        throw new Error('Invalid LiveKit API key format');
      }

      // Validate API secret length - should be a reasonable length
      if (this.apiSecret.length < 32) {
        console.error('[LiveKitService] API secret appears too short');
        throw new Error('Invalid LiveKit API secret - too short');
      }

      const at = new AccessToken(this.apiKey, this.apiSecret, {
        identity: interviewerIdentity,
        name: 'AI Interviewer',
        metadata: JSON.stringify({
          role: 'interviewer',
          config: interviewConfig,
          isBot: true
        })
      });

      // Grant full permissions for the AI interviewer
      at.addGrant({
        roomJoin: true,
        room: roomName,
        canPublish: true,
        canSubscribe: true,
        canPublishData: true,
        canUpdateOwnMetadata: true,
        roomAdmin: true,
      });

      const token = at.toJwt();
      
      // Validate that the token is actually a string
      if (typeof token !== 'string' || token.length === 0) {
        console.error('[LiveKitService] Generated interviewer token is not a valid string:', {
          type: typeof token,
          length: token?.length || 0,
          token: token
        });
        throw new Error(`Invalid interviewer token generated: expected string, got ${typeof token} with length ${token?.length || 0}`);
      }
      
      console.log('[LiveKitService] Interviewer token generated successfully, length:', token.length);
      return token;
    } catch (error) {
      console.error('[LiveKitService] Error generating interviewer token:', error);
      throw new Error(`Failed to generate interviewer token: ${error.message}`);
    }
  }

  /**
   * Create a new interview room
   */
  async createInterviewRoom(interviewConfig, participantName) {
    console.log('[LiveKitService] createInterviewRoom() called');
    console.log('[LiveKitService] Current wsUrl:', `"${this.wsUrl}"`);
    
    const roomName = `interview-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
    
    try {
      // Generate tokens for both participant and AI interviewer
      const participantToken = this.generateAccessToken(
        roomName, 
        participantName,
        {
          role: 'candidate',
          config: interviewConfig,
          joinedAt: new Date().toISOString()
        }
      );

      const interviewerToken = this.generateInterviewerToken(roomName, interviewConfig);

      const roomData = {
        roomName,
        wsUrl: this.wsUrl,
        participantToken,
        interviewerToken,
        config: interviewConfig
      };
      
      console.log('[LiveKitService] Room data created:');
      console.log('- roomName:', roomData.roomName);
      console.log('- wsUrl:', typeof roomData.wsUrl, `"${roomData.wsUrl}"`);
      console.log('- participantToken length:', roomData.participantToken?.length || 0);
      console.log('- interviewerToken length:', roomData.interviewerToken?.length || 0);

      return roomData;
    } catch (error) {
      console.error('[LiveKitService] Error creating interview room:', error);
      throw new Error(`Failed to create interview room: ${error.message}`);
    }
  }

  /**
   * Generate a token for reconnecting to an existing room
   */
  generateReconnectToken(roomName, participantName, metadata = {}) {
    return this.generateAccessToken(roomName, participantName, {
      ...metadata,
      reconnectedAt: new Date().toISOString()
    });
  }

  /**
   * Validate and decode a LiveKit token
   */
  validateToken(token) {
    try {
      const decoded = jwt.verify(token, this.apiSecret);
      return {
        valid: true,
        payload: decoded
      };
    } catch (error) {
      return {
        valid: false,
        error: error.message
      };
    }
  }

  /**
   * Generate webhook verification token
   */
  generateWebhookToken(payload) {
    if (!this.isConfigured()) {
      throw new Error('LiveKit not configured');
    }

    return jwt.sign(payload, this.apiSecret, {
      issuer: this.apiKey,
      expiresIn: '1h'
    });
  }

  /**
   * Verify webhook signature
   */
  verifyWebhookSignature(token, expectedPayload) {
    try {
      const decoded = jwt.verify(token, this.apiSecret, {
        issuer: this.apiKey
      });
      return {
        valid: true,
        payload: decoded
      };
    } catch (error) {
      return {
        valid: false,
        error: error.message
      };
    }
  }
}

File: python_backend/services/livekit_service.py
"""
LiveKit Service for Python backend
"""
import os
from typing import Optional, Dict, Any
from livekit import api
from loguru import logger

class LiveKitService:
    """LiveKit service for managing rooms and tokens"""
    
    def __init__(self):
        self.api_key = os.getenv("LIVEKIT_API_KEY")
        self.api_secret = os.getenv("LIVEKIT_API_SECRET")
        self.ws_url = os.getenv("LIVEKIT_WS_URL")
        
        logger.info("[LiveKitService] Initializing LiveKit service")
        logger.info(f"[LiveKitService] API Key: {'Set' if self.api_key else 'Not set'}")
        logger.info(f"[LiveKitService] API Secret: {'Set' if self.api_secret else 'Not set'}")
        logger.info(f"[LiveKitService] WS URL: {self.ws_url if self.ws_url else 'Not set'}")
        
        if not self.is_configured():
            logger.warning("[LiveKitService] LiveKit credentials not configured properly. Voice interviews will be disabled.")
        else:
            logger.info("[LiveKitService] LiveKit service initialized successfully")
    
    def is_configured(self) -> bool:
        """Check if LiveKit is properly configured"""
        return bool(self.api_key and self.api_secret and self.ws_url)
    
    def get_websocket_url(self) -> Optional[str]:
        """Get the WebSocket URL"""
        return self.ws_url
    
    async def generate_access_token(
        self,
        room_name: str,
        participant_name: str,
        metadata: Optional[Dict[str, Any]] = None
    ) -> str:
        """Generate access token for a participant"""
        if not self.is_configured():
            raise Exception("LiveKit not configured")
        
        logger.info(f"[LiveKitService] Generating access token for {participant_name} in room {room_name}")
        
        try:
            token = api.AccessToken(self.api_key, self.api_secret)
            # Set the participant's identity (required)
            token.with_identity(participant_name)
            # Optionally set a display name
            token.with_name(participant_name)
            
            if metadata:
                token.with_metadata(str(metadata))
            
            # Set room and permissions via VideoGrants
            token.with_grants(api.VideoGrants(
                room_join=True,      # Allow joining the room
                room=room_name,      # Specify the room name (required for agent dispatch)
                can_publish=True,
                can_subscribe=True,
                can_publish_data=True,
                can_update_own_metadata=True
            ))

            # Log for verification
            logger.debug(f"[LiveKitService] AccessToken identity set to: {participant_name}")
            logger.debug(f"[LiveKitService] AccessToken grants: room={room_name}, room_join=True")

            jwt_token = token.to_jwt()
            logger.info(f"[LiveKitService] Token generated successfully, length: {len(jwt_token)}")
            
            return jwt_token
            
        except Exception as error:
            logger.error(f"[LiveKitService] Error generating access token: {error}")
            raise Exception(f"Failed to generate access token: {error}")
    
    async def create_interview_room(
        self,
        interview_config: Dict[str, Any],
        participant_name: str
    ) -> Dict[str, Any]:
        """Create a new interview room"""
        import time
        import random
        import string
        
        room_name = f"interview-{int(time.time())}-{''.join(random.choices(string.ascii_lowercase, k=9))}"
        
        try:
            # Generate tokens for both participant and AI interviewer
            participant_token = await self.generate_access_token(
                room_name,
                participant_name,
                {
                    "role": "candidate",
                    "config": interview_config,
                    "joined_at": "2024-01-01T00:00:00Z"
                }
            )
            
            interviewer_token = await self.generate_access_token(
                room_name,
                f"ai-interviewer-{int(time.time())}",
                {
                    "role": "interviewer",
                    "config": interview_config,
                    "is_bot": True
                }
            )
            
            room_data = {
                "room_name": room_name,
                "ws_url": self.ws_url,
                "participant_token": participant_token,
                "interviewer_token": interviewer_token,
                "config": interview_config
            }
            
            logger.info(f"[LiveKitService] Room created: {room_name}")
            return room_data
            
        except Exception as error:
            logger.error(f"[LiveKitService] Error creating interview room: {error}")
            raise Exception(f"Failed to create interview room: {error}")